README.md
# AutoCanary

AutoCanary is a simple, open source, desktop program for Windows, macOS, and Linux that makes the process of generating machine-readable, digitally signed warrant canary statements simpler.

A warrant canary is a published statement that a service provider has **not** received legal process that it is prohibited from disclosing to the public, such as a national security letter. Once a service provider receives a legal request that contains a gag order, the canary statement is removed. For more information, see [EFF's Warrant Canary FAQ](https://www.eff.org/deeplinks/2014/04/warrant-canary-faq).

Download Windows and macOS X binaries from the [Releases](https://github.com/firstlookmedia/autocanary/releases) page. Linux users can build from source with [these instructions](/BUILD.md).

## How it works

![AutoCanary](/screenshot.png)

Before you begin:

* Choose one person in your organization (probably your General Counsel) to be responsible for signing warrant canary statements. This person must have a PGP key.
* Choose how often you wish to issue canary statements (available options include weekly, monthly, quarterly, and semi-annually).
* Set a recurring event with a reminder in your calendar to sign your canary statement. **This is important:** failing to publish your canary statement on time could result in automated alarms ringing. If your canary stops tweeting, it may lead your users to believe you've received a gag order when you haven't.
* Create a page on your website to publish your warrant canary message, [something like this](https://help.riseup.net/en/canary).

When you run AutoCanary for the first time, first choose how frequently you'd like to publish a warrant canary message. Then choose "All good" for status" -- the other available option, "It's complicated", is there if you need it and are legally allowed to use it. If you do receive a gag order, contact a lawyer to decide how to proceed. Then write your canary statement message. You should customize it to fit your organization, and end with the name of the person who will be signing. Finally, select the PGP key that you'll be using to sign your canary statement.

If you'd like, you can include recent news headlines with each of your canary statements by checking the "Add Recent News Headlines" checkbox. This will prove that the statement was written no earlier than the date you claim it was written.

When you click `Save and Sign` all of these settings will be saved, so the next time you run AutoCanary you won't have to change anything. If you'd wish to sign a warrant canary that's different just this one time and not save your changes, you can click `One-Time Sign` instead.

After clicking a sign button, you may be prompted to enter your PGP passphrase. You'll then see your final digitally-signed warrant canary message. If it looks good to you, post it to your website.

If the person who digitally signs the canary messages knows how to update the website themselves, they can click `Copy to Clipboard`, edit the warrant canary page, and paste it. Otherwise they can click `Save to File`, and then email that file to the person in charge of updating the warrant canary page on the website. *Make sure they update it promptly to prevent the canary from expiring.*

Every time your warrant canary calendar event notifies you, open AutoCanary, generate a new canary message, and update it on your website.

# Installing

### Windows or macOS

For Windows and macOS, download the latest release from the GitHub [releases page](https://github.com/firstlookmedia/autocanary/releases).

### Debian or Ubuntu

Make sure you have `apt-transport-https` installed, and add the FLM code repository key:

```
sudo apt update
sudo apt install -y curl gnupg apt-transport-https
curl -L https://packagecloud.io/firstlookmedia/code/gpgkey | sudo apt-key add -
```

Add the repository, depending on your operating system:

- Ubuntu 18.04 (bionic)
  ```
  echo "deb https://packagecloud.io/firstlookmedia/code/ubuntu/ bionic main" | sudo tee -a /etc/apt/sources.list.d/firstlookmedia_code.list
  ```
- Ubuntu 19.04 (disco)
  ```
  echo "deb https://packagecloud.io/firstlookmedia/code/ubuntu/ disco main" | sudo tee -a /etc/apt/sources.list.d/firstlookmedia_code.list
  ```
- Ubuntu 19.10 (eoan)
  ```
  echo "deb https://packagecloud.io/firstlookmedia/code/ubuntu/ eoan main" | sudo tee -a /etc/apt/sources.list.d/firstlookmedia_code.list
  ```
- Debian 10 (buster)
  ```
  echo "deb https://packagecloud.io/firstlookmedia/code/debian/ buster main" | sudo tee -a /etc/apt/sources.list.d/firstlookmedia_code.list
  ```
- Debian 11 (bullseye)
  ```
  echo "deb https://packagecloud.io/firstlookmedia/code/debian/ bullseye main" | sudo tee -a /etc/apt/sources.list.d/firstlookmedia_code.list
  ```

Install AutoCanary:

```
sudo apt update
sudo apt install -y autocanary
```

### Fedora

We have repositories for Fedora 30 and 31. Add this repository following [these instructions](https://packagecloud.io/firstlookmedia/code/install#manual-rpm), or by running this script:

```
curl -s https://packagecloud.io/install/repositories/firstlookmedia/code/script.rpm.sh | sudo bash
```

Install AutoCanary:

```
sudo dnf install -y autocanary
```


CHANGELOG.md
# Changelog

## 0.2.1

* Packaged for Linux

## 0.2.0

* Optional feature to add recent news headlines to the bottom of canary messages
* Now based on python 3, Qt5, and PyInstaller
* Fixes bug where AutoCanary crashed on older Macs

## 0.1

* First version!


autocanary/__init__.py
"""
AutoCanary | https://firstlook.org/code/autocanary
Copyright (c) 2015 First Look Media

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
"""
import sys, datetime, platform
from PyQt5 import QtCore, QtWidgets

from .gnupg import GnuPG
from .headlines import Headlines
from .settings import Settings
from .output_dialog import OutputDialog

from . import common

class AutoCanaryGui(QtWidgets.QMainWindow):

    def __init__(self, app, gpg, headlines):
        super(AutoCanaryGui, self).__init__()

        self.app = app
        self.gpg = gpg
        self.headlines = headlines
        self.settings = Settings()
        self.setWindowTitle('AutoCanary')
        self.setWindowIcon(common.get_icon())

        # frequency, year
        self.date_col1_layout = QtWidgets.QVBoxLayout()

        self.frequency_layout = QtWidgets.QHBoxLayout()
        self.frequency_label = QtWidgets.QLabel('Frequency')
        self.frequency = QtWidgets.QComboBox()
        frequency_options = ["Weekly", "Monthly", "Quarterly", "Semiannually"]
        for option in frequency_options:
            self.frequency.addItem(option)
        option = self.settings.get_frequency()
        if option in frequency_options:
            self.frequency.setCurrentIndex(frequency_options.index(option))
        self.frequency_layout.addWidget(self.frequency_label)
        self.frequency_layout.addWidget(self.frequency)
        self.frequency.activated.connect(self.update_date)

        self.year_layout = QtWidgets.QHBoxLayout()
        self.year_label = QtWidgets.QLabel('Year')
        self.year = QtWidgets.QComboBox()
        y = datetime.date.today().year
        year_options = [str(y-1), str(y), str(y+1)]
        for option in year_options:
            self.year.addItem(option)
        option = self.settings.get_year()
        if option in year_options:
            self.year.setCurrentIndex(year_options.index(option))
        self.year_layout.addWidget(self.year_label)
        self.year_layout.addWidget(self.year)
        self.year.activated.connect(self.update_date)

        # weekly dropdown
        self.weekly_layout = QtWidgets.QHBoxLayout()
        self.weekly_label = QtWidgets.QLabel('Week')
        self.weekly_dropdown = QtWidgets.QComboBox()
        self.weekly_layout.addWidget(self.weekly_label)
        self.weekly_layout.addWidget(self.weekly_dropdown)

        # monthly dropdown
        self.monthly_layout = QtWidgets.QHBoxLayout()
        self.monthly_label = QtWidgets.QLabel('Month')
        self.monthly_dropdown = QtWidgets.QComboBox()
        monthly_options = ["January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"]
        for option in monthly_options:
            self.monthly_dropdown.addItem(option)
        self.monthly_layout.addWidget(self.monthly_label)
        self.monthly_layout.addWidget(self.monthly_dropdown)

        # quarterly radio buttons
        self.quarterly_layout = QtWidgets.QHBoxLayout()
        self.quarterly_label = QtWidgets.QLabel('Quarter')
        self.quarterly_q1 = QtWidgets.QRadioButton("")
        self.quarterly_q2 = QtWidgets.QRadioButton("")
        self.quarterly_q3 = QtWidgets.QRadioButton("")
        self.quarterly_q4 = QtWidgets.QRadioButton("")
        self.quarterly_layout.addWidget(self.quarterly_label)
        self.quarterly_layout.addWidget(self.quarterly_q1)
        self.quarterly_layout.addWidget(self.quarterly_q2)
        self.quarterly_layout.addWidget(self.quarterly_q3)
        self.quarterly_layout.addWidget(self.quarterly_q4)

        # semiannual radio buttons
        self.semiannually_layout = QtWidgets.QHBoxLayout()
        self.semiannually_label = QtWidgets.QLabel('Semester')
        self.semiannually_q12 = QtWidgets.QRadioButton("")
        self.semiannually_q34 = QtWidgets.QRadioButton("")
        self.semiannually_layout.addWidget(self.semiannually_label)
        self.semiannually_layout.addWidget(self.semiannually_q12)
        self.semiannually_layout.addWidget(self.semiannually_q34)

        # date layout
        self.date_layout = QtWidgets.QVBoxLayout()
        self.date_layout.addLayout(self.frequency_layout)
        self.date_layout.addLayout(self.year_layout)
        self.date_layout.addLayout(self.weekly_layout)
        self.date_layout.addLayout(self.monthly_layout)
        self.date_layout.addLayout(self.quarterly_layout)
        self.date_layout.addLayout(self.semiannually_layout)

        # status
        self.status_layout = QtWidgets.QHBoxLayout()
        self.status_label = QtWidgets.QLabel('Status')
        self.status = QtWidgets.QComboBox()
        status_options = ["All good", "It's complicated"]
        for option in status_options:
            self.status.addItem(option)
        option = self.settings.get_status()
        if option in status_options:
            self.status.setCurrentIndex(status_options.index(option))
        self.status_layout.addWidget(self.status_label)
        self.status_layout.addWidget(self.status)

        # canary text box
        self.textbox = QtWidgets.QTextEdit()
        self.textbox.setText(self.settings.get_text())

        # headlines controls: [checkbox, button, button].
        self.headlines_controls = self.get_headlines_controls()

        # key selection
        seckeys = gpg.seckeys_list()
        self.key_selection = QtWidgets.QComboBox()
        for seckey in seckeys:
            uid = seckey['uid']
            if len(uid) >= 53:
                uid = '{0}...'.format(uid[:50])
            fp = seckey['fp'][-8:]
            text = '{0} [{1}]'.format(uid, fp)
            self.key_selection.addItem(text)
        fp = self.settings.get_fp()
        if fp:
            key_i = 0
            for i, seckey in enumerate(seckeys):
                if seckey['fp'] == fp:
                    key_i = i
            self.key_selection.setCurrentIndex(key_i)

        # buttons
        self.buttons_layout = QtWidgets.QHBoxLayout()
        self.sign_save_button = QtWidgets.QPushButton('Save and Sign')
        self.sign_save_button.clicked.connect(self.sign_save_clicked)
        self.sign_once = QtWidgets.QPushButton('One-Time Sign')
        self.sign_once.clicked.connect(self.sign_once_clicked)
        self.buttons_layout.addWidget(self.sign_save_button)
        self.buttons_layout.addWidget(self.sign_once)

        # headlines controls.
        self.headline_controls_layout = QtWidgets.QHBoxLayout()
        for hl_control in self.headlines_controls:
            self.headline_controls_layout.addWidget(hl_control)

        # layout
        layout = QtWidgets.QVBoxLayout()
        layout.addLayout(self.date_layout)
        layout.addLayout(self.status_layout)
        layout.addWidget(self.textbox)
        layout.addLayout(self.headline_controls_layout)
        layout.addWidget(self.key_selection)
        layout.addLayout(self.buttons_layout)
        central_widget = QtWidgets.QWidget()
        central_widget.setLayout(layout)
        self.setCentralWidget(central_widget)

        self.update_date()
        self.show()

    def update_date(self):
        frequency = self.frequency.currentText()
        year = self.year.currentText()

        if frequency == 'Weekly':
            self.weekly_label.show()
            self.weekly_dropdown.show()

            # regenerate the weekly dropdown options based on the current year
            self.weekly_dropdown.clear()

            one_week = datetime.timedelta(7)
            cur_date = datetime.datetime(int(year), 1, 1)

            def get_monday_of_week(d):
                days_past = d.isoweekday() - 1
                d = d - datetime.timedelta(days_past)
                return d

            def get_sunday_of_week(d):
                days_future = 7 - d.isoweekday()
                d = d + datetime.timedelta(days_future)
                return d

            while cur_date.year == int(year):
                date_str = ''

                # figure out the start and end dates of the week
                monday_date = get_monday_of_week(cur_date)
                sunday_date = get_sunday_of_week(cur_date)

                if sunday_date.year != int(year):
                    break

                date_str += '{0} to {1}'.format(
                    monday_date.strftime('%b %d'),
                    sunday_date.strftime('%b %d')
                )
                self.weekly_dropdown.addItem(date_str)

                cur_date += one_week

        else:
            self.weekly_label.hide()
            self.weekly_dropdown.hide()

        if frequency == 'Monthly':
            self.monthly_label.show()
            self.monthly_dropdown.show()
        else:
            self.monthly_label.hide()
            self.monthly_dropdown.hide()

        if frequency == 'Quarterly':
            self.quarterly_label.show()
            self.quarterly_q1.show()
            self.quarterly_q2.show()
            self.quarterly_q3.show()
            self.quarterly_q4.show()

            # make sure a quarterly radio button is checked
            if not self.quarterly_q1.isChecked() and not self.quarterly_q2.isChecked() and not self.quarterly_q3.isChecked() and not self.quarterly_q4.isChecked():
                self.quarterly_q1.setChecked(True)
        else:
            self.quarterly_label.hide()
            self.quarterly_q1.hide()
            self.quarterly_q2.hide()
            self.quarterly_q3.hide()
            self.quarterly_q4.hide()

        if frequency == 'Semiannually':
            self.semiannually_label.show()
            self.semiannually_q12.show()
            self.semiannually_q34.show()

            # make sure a semiannually radio button is checked
            if not self.semiannually_q12.isChecked() and not self.semiannually_q34.isChecked():
                self.semiannually_q12.setChecked(True)
        else:
            self.semiannually_label.hide()
            self.semiannually_q12.hide()
            self.semiannually_q34.hide()

        # update freqency text

        # the QString objects which represent the widget state are unicode
        # strings, hence the u'...'
        self.quarterly_q1.setText(u'Q1 {}'.format(year))
        self.quarterly_q2.setText(u'Q2 {}'.format(year))
        self.quarterly_q3.setText(u'Q3 {}'.format(year))
        self.quarterly_q4.setText(u'Q4 {}'.format(year))
        self.semiannually_q12.setText(u'Q1 and Q2 {}'.format(year))
        self.semiannually_q34.setText(u'Q3 and Q4 {}'.format(year))

    def get_year_period(self):
        frequency = self.frequency.currentText()

        if frequency == 'Weekly':
            year_period = self.weekly_dropdown.currentText()

        elif frequency == 'Monthly':
            year_period = self.monthly_dropdown.currentText()

        elif frequency == 'Quarterly':
            if self.quarterly_q1.isChecked():
                year_period = 'Q1'
            if self.quarterly_q2.isChecked():
                year_period = 'Q2'
            if self.quarterly_q3.isChecked():
                year_period = 'Q3'
            if self.quarterly_q4.isChecked():
                year_period = 'Q4'

        elif frequency == 'Semiannually':
            if self.semiannually_q12.isChecked():
                year_period = 'Q12'
            if self.semiannually_q34.isChecked():
                year_period = 'Q34'

        return year_period

    def sign_save_clicked(self):
        self.save()
        self.sign()

    def sign_once_clicked(self):
        self.sign()

    def save(self):
        frequency = self.frequency.currentText()
        year = self.year.currentText()
        year_period = self.get_year_period()
        status = str(self.status.currentText())
        text = self.textbox.toPlainText()

        self.settings.set_frequency(frequency)
        self.settings.set_year(year)
        self.settings.set_status(status)
        self.settings.set_text(text)

        key_i = self.key_selection.currentIndex()
        fp = self.gpg.seckeys_list()[key_i]['fp']
        self.settings.set_fp(fp)

        self.settings.save()

    def sign(self):
        frequency = self.frequency.currentText()
        year = self.year.currentText()
        year_period = self.get_year_period()
        status = str(self.status.currentText())
        text = str(self.textbox.toPlainText())

        if self.headlines.enabled and self.headlines.have_headlines:
            text = '\n\n'.join([text, self.headlines.headlines_str, ''])

        # add headers
        period_date = year_period
        if frequency == 'Quarterly':
            quarterly_headers = {
                'Q1': 'January 1 to March 31',
                'Q2': 'April 1 to June 30',
                'Q3': 'July 1 to September 30',
                'Q4': 'October 1 to December 31'
            }
            period_date = quarterly_headers[year_period]
        elif frequency == 'Semiannually':
            semiannually_headers = {
                'Q12': 'January 1 to June 30',
                'Q34': 'July 1 to December 31'
            }
            period_date = semiannually_headers[year_period]

        # the QString objects which represent the widget state are unicode
        # strings, hence the u'...'
        message = u'Status: {}\nPeriod: {}, {}\n\n{}'.format(status, period_date, year, text)

        # sign the file
        key_i = self.key_selection.currentIndex()
        fp = self.gpg.seckeys_list()[key_i]['fp']
        signed_message = self.gpg.sign(message, fp)

        if signed_message:
            # display signed message
            dialog = OutputDialog(self.app, signed_message)
            dialog.exec_()
        else:
            common.alert('Failed to sign message.')

    def get_headlines_controls(self):
        checkbox = QtWidgets.QCheckBox('Add Recent News Headlines')
        button_fetch = QtWidgets.QPushButton('Fetch Headlines')
        button_fetch.setDisabled(True)
        button_preview = QtWidgets.QPushButton('Preview Headlines')
        button_preview.setDisabled(True)
        def fetch_headlines():
            # synchronous.
            self.headlines.fetch_headlines()
            button_fetch.setDisabled(False)
            self.setCursor(QtCore.Qt.ArrowCursor)
            if self.headlines.have_headlines:
                button_preview.setDisabled(False)
        def cb_clicked():
            if checkbox.isChecked():
                button_fetch.setDisabled(False)
                self.headlines.enabled = True
                if self.headlines.have_headlines:
                    button_preview.setDisabled(False)
                else:
                    button_preview.setDisabled(True)
            else:
                self.headlines.enabled = False
                button_fetch.setDisabled(True)
                button_preview.setDisabled(True)
        def fetch_clicked():
            self.setCursor(QtCore.Qt.WaitCursor)
            button_fetch.setDisabled(True)
            # allow the widgets a chance to refresh (disabled/wait).
            QtCore.QTimer().singleShot(1, fetch_headlines)
        def preview_clicked():
            dialog = QtWidgets.QDialog()
            dialog.setModal(True)
            dialog.setWindowFlags(QtCore.Qt.WindowCloseButtonHint)
            layout = QtWidgets.QVBoxLayout()
            text = QtWidgets.QTextEdit()
            text.setText(self.headlines.headlines_str)
            button_close = QtWidgets.QPushButton('Close')
            button_close.clicked.connect(dialog.close)
            layout.addWidget(text)
            layout.addWidget(button_close)
            dialog.setLayout(layout)
            dialog.exec_()
        button_fetch.clicked.connect(fetch_clicked)
        button_preview.clicked.connect(preview_clicked)
        checkbox.clicked.connect(cb_clicked)
        return [checkbox, button_fetch, button_preview]


def main():
    # start the app
    app = QtWidgets.QApplication(sys.argv)

    # initialize and check for gpg and a secret key
    gpg = GnuPG()

    # initialize the module which handles the daily headlines feature.
    headlines = Headlines()

    system = platform.system()
    if system == 'Darwin':
        if not gpg.is_gpg_available():
            common.alert('GPG doesn\'t seem to be installed. Install <a href="https://gpgtools.org/">GPGTools</a>, generate a key, and run AutoCanary again.')
            sys.exit(0)

        seckeys = gpg.seckeys_list()
        if len(seckeys) == 0:
            common.alert('You need an encryption key to use AutoCanary. Run the GPG Keychain program, generate a key, and run AutoCanary again.')
            sys.exit(0)

    elif system == 'Linux':
        seckeys = gpg.seckeys_list()
        if len(seckeys) == 0:
            common.alert('You need an encryption key to use AutoCanary. Generate a key, and run AutoCanary again.')
            sys.exit(0)

    elif system == 'Windows':
        if not gpg.is_gpg_available():
            common.alert('GPG doesn\'t seem to be installed. Install <a href="http://gpg4win.org/">Gpg4win</a>, generate a key, and run AutoCanary again.')
            sys.exit(0)

        seckeys = gpg.seckeys_list()
        if len(seckeys) == 0:
            common.alert('You need an encryption key to use AutoCanary. Run the Kleopatra program, generate a new personal OpenPGP key pair, and run AutoCanary again.')
            sys.exit(0)

    # start the gui
    gui = AutoCanaryGui(app, gpg, headlines)
    sys.exit(app.exec_())


if __name__ == '__main__':
    main()
